### PR TITLE
[Federation] Rewrite ReplicaSet CRUD and Preferences tests.

### PR DESCRIPTION
**Pull request by madhusudancs**
I think `should create replicasets and rebalance them` test is still flaky. I still don't know the source of this flakiness. I will continue hunting. But it is a lot less flaky than before (or perhaps it even never passed before?). This PR could be merged now and flake hunting can happen in parallel.

```release-note
NONE
```

### CODE REVIEW

**Comment by madhusudancs** [comment]
@k8s-bot federation gce e2e test this

File: test/e2e_federation/replicaset.go [hunk 1]
```diff
@@ -462,7 +461,20 @@ func newReplicaSet(namespace string, prefix string, replicas int32, pref *federa
...
+	return rs
+}
+
+func newReplicaSetWithName(namespace string, name string, replicas int32, pref *federation.FederatedReplicaSetPreferences) *v1beta1.ReplicaSet {
+	rs := newReplicaSetObj(namespace, replicas, pref)
+	rs.Name = name
+	return rs
 }
 
 func extraceClusterNames(clusters map[string]*cluster) []string {
```
Comment (perotinus) [inline test/e2e_federation/replicaset.go]: Not your code, but is this supposed to be extractClusterNames?

Comment (madhusudancs) [inline test/e2e_federation/replicaset.go]: Yes! Nice catch. Please send a PR to fix it :)

**Comment by madhusudancs** [comment]
@k8s-bot federation gce e2e test this

File: test/e2e_federation/replicaset.go (hunk 1 above)
Comment (csbell) [inline test/e2e_federation/replicaset.go]: L120: I think you want rs = createReplicaSetOrFail(f.FederationClientset, rs) or else you have the wrong object in the cleanup function that follows.

Comment (madhusudancs) [inline test/e2e_federation/replicaset.go]: Indeed, thanks for catching this. I have another change along with with this to retry on update failure. I want to try out before pushing that. So things are kind of on-hold for now.

**Comment by madhusudancs** [comment]
@k8s-bot federation gce e2e test this

**Comment by madhusudancs** [comment]
@k8s-bot federation gce e2e test this

**Comment by madhusudancs** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by csbell** [comment]
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[test/e2e_federation/replicaset.go, line 110 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42827#-KfYqhjgpYuXeQqG_OpH:-KfYqhjhdrUl6ilConBu:b-9t0upk) ([raw file](https://github.com/kubernetes/kubernetes/blob/e88bab62865e2d2966d8b86684e1dfeda3ac11bf/test/e2e_federation/replicaset.go#L110)):*
> ```Go
> 
> 				rs = newReplicaSet(nsName, FederationReplicaSetPrefix, 5, nil)
> 				rs = createReplicaSetOrFail(f.FederationClientset, rs)
> ```

There's already a createAndWaitForReplicaSetOrFail in the BeforeEach clause. Aren't we clobbering rs within this It()?

---

*[test/e2e_federation/replicaset.go, line 115 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42827#-KfYrAZ6XJkY-4vx8rG9:-KfYrAZ6XJkY-4vx8rGA:b89rkte) ([raw file](https://github.com/kubernetes/kubernetes/blob/e88bab62865e2d2966d8b86684e1dfeda3ac11bf/test/e2e_federation/replicaset.go#L115)):*
> ```Go
> 				By(fmt.Sprintf("Successfully created and synced replicaset \"%s/%s\" (%v/%v) to clusters", nsName, rs.Name, rs.Status.Replicas, *rs.Spec.Replicas))
> 
> 				rs = newReplicaSetWithName(nsName, rs.Name, 15, nil)
> ```

I'd like to see some mention in comments or It() that we're actually scaling the replicaset.

---

*[test/e2e_federation/replicaset.go, line 180 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42827#-KfYsCCeopcIB0q3A5t8:-KfYsCCfwOxRnMXsZ701:bwr6wk4) ([raw file](https://github.com/kubernetes/kubernetes/blob/e88bab62865e2d2966d8b86684e1dfeda3ac11bf/test/e2e_federation/replicaset.go#L180)):*
> ```Go
> 			It("should create replicasets and rebalance them", func() {
> 				nsName := f.FederationNamespace.Name
> 				pref1, pref2, replicas, expect1, expect2 := generageFedRsPrefsForRebalancing(clusters)
> ```

generage -> generate?

---

*[test/e2e_federation/replicaset.go, line 363 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42827#-KfYsMGlGGr4KjJ1iR9M:-KfYsMGlGGr4KjJ1iR9N:bvj10ws) ([raw file](https://github.com/kubernetes/kubernetes/blob/e88bab62865e2d2966d8b86684e1dfeda3ac11bf/test/e2e_federation/replicaset.go#L363)):*
> ```Go
> func generageFedRsPrefsForRebalancing(clusters map[string]*cluster) (pref1, pref2 *federation.FederatedReplicaSetPreferences, replicas int32, expect1, expect2 map[string]int32) {
> 	By("Generating replicaset for rebalancing")
> 	clusterNames := extraceClusterNames(clusters)
> ```

I think the extrace was fixed somewhere else... although maybe not 1.6

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42827)*

**Comment by madhusudancs** [comment]
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[test/e2e_federation/replicaset.go, line 110 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42827#-KfYqhjgpYuXeQqG_OpH:-KfZVKd9nC0NTSBXWdAz:b-n9tnyf) ([raw file](https://github.com/kubernetes/kubernetes/blob/e88bab62865e2d2966d8b86684e1dfeda3ac11bf/test/e2e_federation/replicaset.go#L110)):*
<details><summary><i>Previously, csbell (Christian Bell) wrote…</i></summary><blockquote>

There's already a createAndWaitForReplicaSetOrFail in the BeforeEach clause. Aren't we clobbering rs within this It()?
</blockquote></details>

You are right. This clobbers the one created by `BeforeEach()`. Fixed.

---

*[test/e2e_federation/replicaset.go, line 115 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42827#-KfYrAZ6XJkY-4vx8rG9:-KfZWu6DzNodcsjy0bzX:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/e88bab62865e2d2966d8b86684e1dfeda3ac11bf/test/e2e_federation/replicaset.go#L115)):*
<details><summary><i>Previously, csbell (Christian Bell) wrote…</i></summary><blockquote>

I'd like to see some mention in comments or It() that we're actually scaling the replicaset.
</blockquote></details>

Done.

---

*[test/e2e_federation/replicaset.go, line 180 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42827#-KfYsCCeopcIB0q3A5t8:-KfZVXn2rYJfhEeTmp6B:bvt8nf0) ([raw file](https://github.com/kubernetes/kubernetes/blob/e88bab62865e2d2966d8b86684e1dfeda3ac11bf/test/e2e_federation/replicaset.go#L180)):*
<details><summary><i>Previously, csbell (Christian Bell) wrote…</i></summary><blockquote>

generage -> generate?
</blockquote></details>

Missed a spot :) Fixed.

---

*[test/e2e_federation/replicaset.go, line 363 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42827#-KfYsMGlGGr4KjJ1iR9M:-KfZVgH83urcsvVK7uk-:b-scrlhs) ([raw file](https://github.com/kubernetes/kubernetes/blob/e88bab62865e2d2966d8b86684e1dfeda3ac11bf/test/e2e_federation/replicaset.go#L363)):*
<details><summary><i>Previously, csbell (Christian Bell) wrote…</i></summary><blockquote>

I think the extrace was fixed somewhere else... although maybe not 1.6
</blockquote></details>

@perotinus' PR is here https://github.com/kubernetes/kubernetes/pull/43035 and is waiting for the gates to open. I did not want to stomp on it.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42827)*

**Pull request merged**